Feature/real time notifications

vic2019 opened the pull request:
- [ ] socketService
- [ ] SocketClient component

vic2019 committed:
Create socketService.js, Socketclient component, Notification model

vic2019 committed:
Remove Notification.js. Add comments to SocketClient and socketService

vic2019 committed:
Merge branch 'dev' into feature/real-time-notifications

aeciorc commented on server/app/socketService.js:
```diff
@@ -0,0 +1,103 @@
+const ENDPOINT = process.env.SOCKET_CHANNEL_ENDPOINT || 'http://localhost';
+const PORT = process.env.SOCKET_CHANNEL_PORT || 3002;
+
+const io = SocketIo();
```
You can use the same server for both, actually. It would be smth like this:
```
const app = require("./app");
const http = require("http").Server(app);
const io = require("socket.io")(http);
...
const server = http.listen(
  PORT,
  console.log(
    `Server running in ${process.env.NODE_ENV} mode on port ${PORT}`
  )
);
```

aeciorc replied:
(on server.js)

aeciorc commented on server/app/socketService.js:
```diff
@@ -0,0 +1,103 @@
+  };
+
+  return admin;
+};
```
I don't think you need the admin client. You can use `io.to(socketId).emit()`

aeciorc commented on server/app/socketService.js:
```diff
@@ -0,0 +1,103 @@
+
+  // A separate endpoint for triggering the emittance of 'notification' event
+  const messageChannel = SocketIo(
+    httpServer.listen(PORT, err => {
```
seems like this isn't actually used, because you attach the other server

vic2019 replied:
This server is for receiving messages from the web scraping service

maismin commented on server/app/socketService.js:
```diff
@@ -0,0 +1,103 @@
+        }
+      });
+    });
+  });
```
Instead of passing the user's email, we can make use of the token in the cookie. I think it's more secure as the token cannot be accessed via JavaScript on the client.
```
    const cookie = require('cookie');
    const jwt = require('jsonwebtoken');
    ...
    // Use the token cookie to authenticate the client
    socket.on('authenticate', async () => {
      const { token } = cookie.parse(socket.handshake.headers.cookie);

      if (!token) {
        return socket.disconnect(true);
      }

      try {
        const decoded = jwt.verify(token, process.env.JWT_SECRET);
        const { id } = decoded;

        await User.findById(id);

        // eslint-disable-next-line no-param-reassign
        socket.authenticated = true;

        // Add current connection to list of connected clients
        socketServer.clients[id] = socket;

        if (process.env.NODE_ENV !== 'production') {
          console.log(Object.keys(socketServer.clients));
        }

        socket.on('disconnect', () => {
          delete socketServer.clients[id];
        });
      } catch (error) {
        console.log(error);
        return socket.disconnect(true);
      }
    });
```

aeciorc replied:
I think this can be done on the 'connect' event listener as well

maismin commented on server/app/socketService.js:
```diff
@@ -0,0 +1,103 @@
+
+    setTimeout(() => {
+      if (!socket.authenticated) return socket.disconnect(true);
+    }, 2000);
```
Is there a case where this might be needed? Does the 'authenticate' listener not take care of this?

aeciorc replied:
Yes, shouldn't be necessary

maismin committed:
Fix S3 upload

- Move the S3 upload to be a service instead of an API route for the
backend. This is because the upload service is used by two other routes,
namely the user's profile image and shopping list cover.

maismin committed:
Add API route to update user's profile image

maismin committed:
Fix Profile component to show user's profile image

- The default profile image will be shown if the user's photoUrl
property is an empty string.

maismin committed:
Fix Documentation for Update User's Profile Image

maismin committed:
Remove unnecessary console logs

maismin committed:
Add functionality to change user's profile image

- Add action types, asynchronous actions, and updates to auth reducer to
update user's profile image.

- Fix Profile component so that when the user clicks on the profile
image and selects another image, that image becomes the new profile image.

maismin committed:
Refactor add profile image functionality

- Refactor auth reducer to use object spread operator instead of
Object.assign to create new objects with new or updated values.

- Refactor the propType of the Profile component to group the id and photo
under the User object instead of being separate as they are logically
part of the User.

maismin committed:
Merge branch 'dev' into feature/add-profile-image

- This refactors the api/selectors into the error and loading reducers.

maismin committed:
Fix Profile component

- The component now imports the createLoadingSelector correctly.

maismin committed:
Refactor Profile component

- Move ProfileImage portion to its own component from the Profile
component.

Esodey reviewed (approved):
After changing Aceio's and Simon's comments, it looks good to go

maismin committed:
Add Dialog component to upload profile image

- Clicking on the profile image now opens a dialog for the user to
upload a profile image. The image size limit is 1 MB.

maismin committed:
Move services folder to within the /app folder

maismin committed:
Merge branch 'dev' into feature/real-time-notifications

maismin committed:
Refactor socket service script

- Move the socketService.js script to /services.

- The token cookie is used to verify the user instead of email.

- Install cookie dependency to read cookies from the socketService
script.

- Refactor SocketClient component to grab isAuthenticated instead of
user's email from the Redux store.
